### PR TITLE
Bring ush/rocoto from py2 to py3

### LINKED ISSUE #265: Convert setup scripts to python 3
Setup scripts are currently python 2, which is at end of life. Convert setup scripts to python 3.

Scripts under ush/rocoto.

### PR DESCRIPTION
**Pull request by BrianCurtis-NOAA**
Bring the ush/rocoto python2 code to python3 standards.

Closes #265

### CODE REVIEW

**Comment by aerorahul** [comment]
@BrianCurtis-NOAA 
Please leave the `rocoto_viewer.py` out of this PR.  Save your version in some branch for later review.
Let this PR be of just the setup scripts.

**Comment by BrianCurtis-NOAA** [comment]
> 
> 
> @BrianCurtis-NOAA
> Please leave the `rocoto_viewer.py` out of this PR. Save your version in some branch for later review.
> Let this PR be of just the setup scripts.

Done

**Comment by KateFriedman-NOAA** [comment]
@BrianCurtis-NOAA I tested both setup script pairs (on Venus) to check they can successfully create an EXPDIR and xml.

- cycled (setup_expt.py and setup_workflow.py): EXPDIR and xml generation worked
- free-forecast (setup_expt_fcstonly.py and setup_workflow_fcstonly.py): error overwriting old EXPDIR of same name from setup_expt_fcstonly.py, I did not encounter same error with cycled versions in similar test, see error below, I have not yet tried setup_workflow_fcstonly.py, thanks!

```
[Kate.Friedman@v71a1 rocoto]$ which python
/usrx/local/prod/packages/python/3.6.3/bin/python
[Kate.Friedman@v71a1 rocoto]$ pwd
/gpfs/dell2/emc/modeling/save/Kate.Friedman/git/global-workflow/feature-rocoto-py3/ush/rocoto
[Kate.Friedman@v71a1 rocoto]$ ./setup_expt_fcstonly.py --pslot testff --res 768 --comrot /gpfs/dell3/ptmp/Kate.Friedman/comrot --expdir /gpfs/dell2/emc/modeling/save/Kate.Friedman/expdir --configdir /gpfs/dell2/emc/modeling/save/Kate.Friedman/git/global-workflow/feature-rocoto-py3/parm/config --gfs_cyc 4 --idate 2021071900 --edate 2021071906

EXPDIR already exists in /gpfs/dell2/emc/modeling/save/Kate.Friedman/expdir/testff

Traceback (most recent call last):
  File "./setup_expt_fcstonly.py", line 248, in <module>
    overwrite_expdir = raw_input('Do you wish to over-write EXPDIR [y/N]: ')
NameError: name 'raw_input' is not defined
```

**Comment by BrianCurtis-NOAA** [comment]
@KateFriedman-NOAA Thanks! Fixed

**Comment by KateFriedman-NOAA** [comment]
@BrianCurtis-NOAA Thanks for the fix! I tested both pairs on Orion and they worked. Overwriting existing EXPDIRs and COMROTs also worked.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by KateFriedman-NOAA** [comment]
Ran both the develop branch and feature/rocoto-py3 setup scripts and compared their resulting EXPDIRs and xmls. See setup steps down below and diffs before that. The only diffs are expected due to different PSLOTs. Looks good! Thanks @BrianCurtis-NOAA !

free-forecast diffs:
```
Orion-login-1[243] /work/noaa/global/kfriedma/expdir$ diff -r testffdev testffpy3
diff -r testffdev/config.base testffpy3/config.base
24c24
< export HOMEgfs=/work/noaa/global/kfriedma/git/develop
---
> export HOMEgfs=/work/noaa/global/kfriedma/git/feature-rocoto-py3
101c101
< export PSLOT="testffdev"
---
> export PSLOT="testffpy3"
Only in testffdev: testffdev.crontab
Only in testffdev: testffdev.xml
Only in testffpy3: testffpy3.crontab
Only in testffpy3: testffpy3.xml
Orion-login-1[245] /work/noaa/global/kfriedma/expdir$ diff testffdev/testffdev.xml testffpy3/testffpy3.xml
13c13
<               This workflow was automatically generated at 2021-07-20 10:18:19.947886
---
>               This workflow was automatically generated at 2021-07-20 10:23:12.399810
17c17
<       <!ENTITY PSLOT    "testffdev">
---
>       <!ENTITY PSLOT    "testffpy3">
30,31c30,31
<       <!ENTITY EXPDIR "/work/noaa/global/kfriedma/expdir/testffdev">
<       <!ENTITY ROTDIR "/work/noaa/stmp/kfriedma/comrot/testffdev">
---
>       <!ENTITY EXPDIR "/work/noaa/global/kfriedma/expdir/testffpy3">
>       <!ENTITY ROTDIR "/work/noaa/stmp/kfriedma/comrot/testffpy3">
35,36c35,36
<       <!ENTITY HOMEgfs  "/work/noaa/global/kfriedma/git/develop">
<       <!ENTITY JOBS_DIR "/work/noaa/global/kfriedma/git/develop/jobs/rocoto">
---
>       <!ENTITY HOMEgfs  "/work/noaa/global/kfriedma/git/feature-rocoto-py3">
>       <!ENTITY JOBS_DIR "/work/noaa/global/kfriedma/git/feature-rocoto-py3/jobs/rocoto">
```
cycled diffs:
```
Orion-login-1[244] /work/noaa/global/kfriedma/expdir$ diff -r testcycdev testcycpy3
diff -r testcycdev/config.base testcycpy3/config.base
24c24
< export HOMEgfs=/work/noaa/global/kfriedma/git/develop
---
> export HOMEgfs=/work/noaa/global/kfriedma/git/feature-rocoto-py3
101c101
< export PSLOT="testcycdev"
---
> export PSLOT="testcycpy3"
Only in testcycdev: testcycdev.crontab
Only in testcycdev: testcycdev.xml
Only in testcycpy3: testcycpy3.crontab
Only in testcycpy3: testcycpy3.xml
Orion-login-1[246] /work/noaa/global/kfriedma/expdir$ diff testcycdev/testcycdev.xml testcycpy3/testcycpy3.xml
13c13
<               This workflow was automatically generated at 2021-07-20 10:19:01.262319
---
>               This workflow was automatically generated at 2021-07-20 10:23:58.861122
17c17
<       <!ENTITY PSLOT "testcycdev">
---
>       <!ENTITY PSLOT "testcycpy3">
30,31c30,31
<       <!ENTITY EXPDIR "/work/noaa/global/kfriedma/expdir/testcycdev">
<       <!ENTITY ROTDIR "/work/noaa/stmp/kfriedma/comrot/testcycdev">
---
>       <!ENTITY EXPDIR "/work/noaa/global/kfriedma/expdir/testcycpy3">
>       <!ENTITY ROTDIR "/work/noaa/stmp/kfriedma/comrot/testcycpy3">
34,35c34,35
<       <!ENTITY HOMEgfs  "/work/noaa/global/kfriedma/git/develop">
<       <!ENTITY JOBS_DIR "/work/noaa/global/kfriedma/git/develop/jobs/rocoto">
---
>       <!ENTITY HOMEgfs  "/work/noaa/global/kfriedma/git/feature-rocoto-py3">
>       <!ENTITY JOBS_DIR "/work/noaa/global/kfriedma/git/feature-rocoto-py3/jobs/rocoto">
```

Setting up develop EXPDIRs:
free-forecast:
```
Orion-login-1[226] /work/noaa/global/kfriedma/git/develop/ush/rocoto$ ./setup_expt_fcstonly.py --pslot testffdev --res 192 --comrot /work/noaa/stmp/kfriedma/comrot --expdir /work/noaa/global/kfriedma/expdir --idate 2020090118 --edate 2020090200 --configdir /work/noaa/global/kfriedma/git/develop/parm/config --gfs_cyc 4

SDATE = 2020-09-01 18:00:00
EDATE = 2020-09-02 00:00:00

EDITED:  /work/noaa/global/kfriedma/expdir/testffdev/config.base as per user input.
DEFAULT: /work/noaa/global/kfriedma/expdir/testffdev/config.base.default is for reference only.
Please verify and delete the default file before proceeding.

Orion-login-1[227] /work/noaa/global/kfriedma/git/develop/ush/rocoto$ ./setup_workflow_fcstonly.py --expdir /work/noaa/global/kfriedma/expdir/testffdev
sourcing config.getic
sourcing config.init
sourcing config.waveinit
sourcing config.waveprep
sourcing config.fcst
sourcing config.post
sourcing config.wavepostsbs
sourcing config.wavepostbndpnt
sourcing config.wavepostbndpntbll
sourcing config.wavepostpnt
sourcing config.wavegempak
sourcing config.waveawipsbulls
sourcing config.waveawipsgridded
sourcing config.wafs
sourcing config.wafsgrib2
sourcing config.wafsblending
sourcing config.wafsgcip
sourcing config.wafsgrib20p25
sourcing config.wafsblending0p25
sourcing config.postsnd
sourcing config.gempak
sourcing config.awips
sourcing config.vrfy
sourcing config.metp
sourcing config.arch
```
cycled:
```
Orion-login-1[228] /work/noaa/global/kfriedma/git/develop/ush/rocoto$ ./setup_expt.py --pslot testcycdev --resdet 192 --resens 96 --comrot /work/noaa/stmp/kfriedma/comrot --expdir /work/noaa/global/kfriedma/expdir --idate 2020090118 --edate 2020090200 --configdir /work/noaa/global/kfriedma/git/develop/parm/config --nens 80 --gfs_cyc 4

SDATE = 2020-09-01 18:00:00
EDATE = 2020-09-02 00:00:00

EDITED:  /work/noaa/global/kfriedma/expdir/testcycdev/config.base as per user input.
DEFAULT: /work/noaa/global/kfriedma/expdir/testcycdev/config.base.emc.dyn is for reference only.
Please verify and delete the default file before proceeding.

Orion-login-1[229] /work/noaa/global/kfriedma/git/develop/ush/rocoto$ ./setup_workflow.py --expdir /work/noaa/global/kfriedma/expdir/testcycdev
sourcing config.prep
sourcing config.anal
sourcing config.analdiag
sourcing config.analcalc
sourcing config.gldas
sourcing config.fcst
sourcing config.postsnd
sourcing config.post
sourcing config.vrfy
sourcing config.arch
sourcing config.eobs
sourcing config.ediag
sourcing config.eomg
sourcing config.eupd
sourcing config.ecen
sourcing config.esfc
sourcing config.efcs
sourcing config.echgres
sourcing config.epos
sourcing config.earc
sourcing config.metp
```

Setting up EXPDIRs with feature/rocoto-py3:
free-forecast:
```
Orion-login-1[235] /work/noaa/global/kfriedma/git/feature-rocoto-py3/ush/rocoto$ module load python/3.7.5
Orion-login-1[236] /work/noaa/global/kfriedma/git/feature-rocoto-py3/ush/rocoto$ module list

Currently Loaded Modules:
  1) munge/0.5.13      3) contrib/0.1     5) rocoto/1.3.3   7) intel/2018.4   9) hpc-intel/2018.4  11) hpc-impi/2018.4  13) python/3.7.5
  2) slurm/19.05.3-2   4) noaatools/1.0   6) hpc/1.1.0      8) mkl/2020.2    10) impi/2018.4       12) prod_util/1.2.2



Orion-login-1[237] /work/noaa/global/kfriedma/git/feature-rocoto-py3/ush/rocoto$ ./setup_expt_fcstonly.py --pslot testffpy3 --res 192 --comrot /work/noaa/stmp/kfriedma/comrot --expdir /work/noaa/global/kfriedma/expdir --idate 2020090118 --edate 2020090200 --gfs_cyc 4 --configdir /work/noaa/global/kfriedma/git/feature-rocoto-py3/parm/config

SDATE = 2020-09-01 18:00:00
EDATE = 2020-09-02 00:00:00

EDITED:  /work/noaa/global/kfriedma/expdir/testffpy3/config.base as per user input.
DEFAULT: /work/noaa/global/kfriedma/expdir/testffpy3/config.base.default is for reference only.
Please verify and delete the default file before proceeding.

Orion-login-1[238] /work/noaa/global/kfriedma/git/feature-rocoto-py3/ush/rocoto$ ./setup_workflow_fcstonly.py --expdir /work/noaa/global/kfriedma/expdir/testffpy3
sourcing config.getic
sourcing config.init
sourcing config.waveinit
sourcing config.waveprep
sourcing config.fcst
sourcing config.post
sourcing config.wavepostsbs
sourcing config.wavepostbndpnt
sourcing config.wavepostbndpntbll
sourcing config.wavepostpnt
sourcing config.wavegempak
sourcing config.waveawipsbulls
sourcing config.waveawipsgridded
sourcing config.wafs
sourcing config.wafsgrib2
sourcing config.wafsblending
sourcing config.wafsgcip
sourcing config.wafsgrib20p25
sourcing config.wafsblending0p25
sourcing config.postsnd
sourcing config.gempak
sourcing config.awips
sourcing config.vrfy
sourcing config.metp
sourcing config.arch
```
cycled:
```
Orion-login-1[239] /work/noaa/global/kfriedma/git/feature-rocoto-py3/ush/rocoto$ ./setup_expt.py --pslot testcycpy3 --resdet 192 --resens 96 --comrot /work/noaa/stmp/kfriedma/comrot --expdir /work/noaa/global/kfriedma/expdir --idate 2020090118 --edate 2020090200 --nens 80 --gfs_cyc 4 --configdir /work/noaa/global/kfriedma/git/feature-rocoto-py3/parm/config

SDATE = 2020-09-01 18:00:00
EDATE = 2020-09-02 00:00:00

EDITED:  /work/noaa/global/kfriedma/expdir/testcycpy3/config.base as per user input.
DEFAULT: /work/noaa/global/kfriedma/expdir/testcycpy3/config.base.emc.dyn is for reference only.
Please verify and delete the default file before proceeding.

Orion-login-1[240] /work/noaa/global/kfriedma/git/feature-rocoto-py3/ush/rocoto$ ./setup_workflow.py --expdir /work/noaa/global/kfriedma/expdir/testcycpy3
sourcing config.prep
sourcing config.anal
sourcing config.analdiag
sourcing config.analcalc
sourcing config.gldas
sourcing config.fcst
sourcing config.postsnd
sourcing config.post
sourcing config.vrfy
sourcing config.arch
sourcing config.eobs
sourcing config.ediag
sourcing config.eomg
sourcing config.eupd
sourcing config.ecen
sourcing config.esfc
sourcing config.efcs
sourcing config.echgres
sourcing config.epos
sourcing config.earc
sourcing config.metp
```

**Review by KateFriedman-NOAA — APPROVED** [review]
Based on tests of setup scripts and comparison of resulting EXPDIRs/xmls these changes look good. Thanks for converting these scripts @BrianCurtis-NOAA !

**Review by aerorahul — APPROVED** [review]
I did as thorough look as I could given the extent of the changes.
It looks good to the eye.
@KateFriedman-NOAA confirmed the outputs are equivalent.

File: ush/rocoto/workflow_utils.py [hunk 1]
```diff
@@ -275,27 +275,33 @@ def get_resources(machine, cfg, task, reservation, cdump='gdas'):
...
-    if cdump in ['gfs'] and 'npe_%s_gfs' % task in cfg.keys():
-        tasks = cfg['npe_%s_gfs' % ltask]
+    if cdump in ['gfs'] and f'npe_{task}_gfs' in cfg.keys():
+        tasks = cfg[f'npe_{ltask}_gfs']
     else:
-        tasks = cfg['npe_%s' % ltask]
+        try:
+            tasks = cfg[f'npe_{ltask}']
+        except KeyError:
+            tasks = cfg["',)npe_waveawipsgridded"]
```
Comment (aerorahul) [inline ush/rocoto/workflow_utils.py:293]: @BrianCurtis-NOAA 
What is this try block supposed to do?
If there is a `KeyError`, what is the key `',)npewaveawipsgridded`?
My guess is the `except KeyError` will not be encountered.

File: ush/rocoto/workflow_utils.py [hunk 2]
```diff
@@ -275,27 +275,33 @@ def get_resources(machine, cfg, task, reservation, cdump='gdas'):
...
+        ppn = cfg[f'npe_node_{ltask}_gfs']
+    else:        
+        ppn = cfg[f'npe_node_{ltask}']
 
     if machine in [ 'WCOSS_DELL_P3', 'HERA', 'ORION']:
-        threads = cfg['nth_%s' % ltask]
+        try:
+            threads = cfg[f'nth_{ltask}']
+        except KeyError:
+            threads = cfg["',)nth_epos"]
```
Comment (aerorahul) [inline ush/rocoto/workflow_utils.py:304]: @BrianCurtis-NOAA 
Same comment as above.
I don't think we have a key `',)nth_epos` in the `cfg` dictionary.

**Comment by BrianCurtis-NOAA** [comment]
@aerorahul
```
'/scratch1/NCEPDEV/global/glopara/git/verif/global/tags/vsdb', 'INCVARS_ZERO_STRAT': "'sphum_inc','liq_wat_inc','icmr_inc'", 'IAUFHRS_ENKF': '3,6,9', "',)nth_epos": 6, 'NEMSIO_IN': '.false.', 'SENDSDM': 'NO', 'LOGSCRIPT': '',
```
There is a key (or two) in the cfg dictionary that is written with `',)` in it, and I don't know why.

**Pull request merged**